src/util/taproot.rs

```diff
@@ -82,12 +93,776 @@ sha256t_hash_newtype!(TapSighashHash, TapSighashTag, MIDSTATE_TAPSIGHASH, 64,
+    pub fn from_u8(ver: u8) -> Result<Self, TaprootError> {
+        if ver & TAPROOT_LEAF_MASK != ver {
+            Err(TaprootError::InvalidTaprootLeafVersion(ver))
+        } else {
+            Ok(LeafVersion { ver: ver })
+        }
+    }
+
+    /// Obtain a LeafVersion from u8 after applying TAPROOT_LEAF_MASK
+    pub fn from_unmasked_u8(ver: u8) -> Self {
+        LeafVersion { ver: ver & TAPROOT_LEAF_MASK }
+    }
+
+    /// Get the inner version from LeafVersion
+    pub fn as_u8(&self) -> u8 {
+        self.ver
+    }
+}
+
+impl Into<u8> for LeafVersion {
+    fn into(self) -> u8 {
+        self.ver
+    }
+}
+
+/// Detailed error type for taproot builder
+#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub enum TaprootBuilderError {
+    /// Merkle Tree depth must not be more than 128
+    InvalidMerkleTreeDepth(usize),
+    /// Nodes must be added specified in DFS order
+    NodeNotInDfsOrder,
+    /// Two nodes at depth 0 are not allowed
+    OverCompleteTree,
+    /// Invalid taproot internal key
+    InvalidInternalKey(secp256k1::Error),
+    /// Called finalize on an incomplete tree
+    IncompleteTree,
+    /// Called finalize on a empty tree
+    EmptyTree,
+}
+
+impl fmt::Display for TaprootBuilderError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match *self {
+            TaprootBuilderError::NodeNotInDfsOrder => {
+                write!(f, "add_leaf/add_hidden must be called in DFS walk order",)
+            }
+            TaprootBuilderError::OverCompleteTree => write!(
+                f,
+                "Attempted to create a tree with two nodes at depth 0. There must\
+                only be a exactly one node at depth 0",
+            ),
+            TaprootBuilderError::InvalidMerkleTreeDepth(d) => write!(
+                f,
+                "Merkle Tree depth({}) must be less than {}",
+                d, TAPROOT_CONTROL_MAX_NODE_COUNT
+            ),
+            TaprootBuilderError::InvalidInternalKey(e) => {
+                write!(f, "Invalid Internal XOnly key : {}", e)
+            }
+            TaprootBuilderError::IncompleteTree => {
+                write!(f, "Called finalize on an incomplete tree")
+            }
+            TaprootBuilderError::EmptyTree => {
+                write!(f, "Called finalize on an empty tree")
+            }
+        }
+    }
+}
+
+/// Detailed error type for taproot utilities
+#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub enum TaprootError {
+    /// Proof size must be a multiple of 32
+    InvalidMerkleBranchSize(usize),
+    /// Merkle Tree depth must not be more than 128
+    InvalidMerkleTreeDepth(usize),
+    /// The last bit of tapleaf version must be zero
+    InvalidTaprootLeafVersion(u8),
+    /// Invalid Control Block Size
+    InvalidControlBlockSize(usize),
+    /// Invalid taproot internal key
+    InvalidInternalKey(secp256k1::Error),
```

Comment:
We need either `source` on this type - or just skip the inner error since it will always be `secp256k1::Error::InvalidSecretKey` without adding any useful information – but making enumeration of error types by library users more complex.

Reply:
The idea is to add `source` when we have it in MSRV. Since the MSRV change is expected soon, I think it's okay to have it is right now. I have added a TODO comment stating the same.